### PR TITLE
Deprecate public API for hash table creation

### PR DESCRIPTION
**Pull request by kgaillot**
This is part of an ongoing project to deprecate generic parts of the public C API so the Pacemaker libraries focus on Pacemaker-specific APIs, giving us greater freedom to reimplement generic utilities as needed.

### CODE REVIEW

**Comment by kgaillot** [comment]
@clumens, can you review this when you get a chance? Thanks

**Comment by clumens** [comment]
Are we likely to have API for making other kinds of tables?  If not, pcmk__str_table_new could just be pcmk__table_new.

**Comment by clumens** [comment]
This looks good to me.

**Comment by kgaillot** [comment]
> Are we likely to have API for making other kinds of tables? If not, pcmk__str_table_new could just be pcmk__table_new.

That's a good question, I was wondering that myself. Especially since "str_table" previously referred to both the key and value being strings, and now it's just the key that has to be a string.

We have one table created with g_str_hash() instead of pcmk__str_hash() (introduced in 2012, likely not for any particular reason). That table is internal to libcrmservice (not affecting scheduler regression tests or anything like that), so changing it to the new function shouldn't cause any issues.

Then we have 6 tables that use g_direct_hash(). Most of those use GINT_TO_POINTER() for the key; we could create convenience functions for them. A couple use a struct pointer, so those should continue using glib directly.

Summing up:
* new hash table with case-sensitive string key (most common case)
* new hash table with case-insensitive string key
* new hash table with integer key
* insert into hash table with integer key

Naming suggestions welcome. As an aside, we're inconsistent about whether new goes at the beginning or end of constructors. Some possibilities:

    // Simple and direct
    table = pcmk__new_table(free, free); // case-sensitive string as default
    table = pcmk__new_strcase_table(free, free); // or pcmk__new_casei_table()
    table = pcmk__new_int_table(free); // Integer keys don't need to be freed

    // Most similar to glib and what we have now
    table = pcmk__str_table_new(free, free);
    table = pcmk__int_table_new(free);

    // Clear and succinct
    table = pcmk__strkey_table(free, free);
    table = pcmk__intkey_table(free);

    // Clearest but getting long
    table = pcmk__new_strkey_table(free, free);
    table = pcmk__new_strcasekey_table(free, free);

**Comment by clumens** [comment]
> > Are we likely to have API for making other kinds of tables? If not, pcmk__str_table_new could just be pcmk__table_new.

Incidentally, I deleted that comment a minute or two later when I saw there was a strcase version as well.  Heh.  I'll take a deeper look at the rest of your reply here in just a minute.

**Comment by clumens** [comment]
> Naming suggestions welcome. As an aside, we're inconsistent about whether new goes at the beginning or end of constructors.

Either way works for me.  I like putting new at the beginning because it reads more naturally left-to-right, but I also like putting it at the end because it means I know what type of thing I'm looking at before I know what operation I'm applying to that type.  So, whichever you go with is fine.

> ```
> // Simple and direct
> table = pcmk__new_table(free, free); // case-sensitive string as default
> table = pcmk__new_strcase_table(free, free); // or pcmk__new_casei_table()
> table = pcmk__new_int_table(free); // Integer keys don't need to be freed
> 
> // Most similar to glib and what we have now
> table = pcmk__str_table_new(free, free);
> table = pcmk__int_table_new(free);
> 
> // Clear and succinct
> table = pcmk__strkey_table(free, free);
> table = pcmk__intkey_table(free);
> 
> // Clearest but getting long
> table = pcmk__new_strkey_table(free, free);
> table = pcmk__new_strcasekey_table(free, free);
> ```

Even though it doesn't include the operation in it, I kind of like clear and succinct.  The long ones are definitely pretty long, and we use a lot of long names in a lot of other places.  It can be a little much.  I  guess I don't have any strong preferences here.  I just wanted to be prepared for the possibility of having multiple types of hash tables that we need to manage.

**Comment by kgaillot** [comment]
Updated crm_str_table_new() and crm_strcase_table_new() commits, and added a commit for integer keys

I went with pcmk__strkey_table(), pcmk__strikey_table(), and pcmk__intkey_table(). I considered pcmk__strcasekey_table() and pcmk__cistrkey_table() for case-insensitive but went with strikey because it's easier on the eyes and a little funny.

File: include/crm/common/util_compat.h [hunk 1]
```diff
@@ -69,6 +69,9 @@ char *pcmk_format_nvpair(const char *name, const char *value,
 //! \deprecated Use a standard printf()-style function instead
 char *pcmk_format_named_time(const char *name, time_t epoch_time);
 
+//! \deprecated Use g_str_hash() instead
+#define crm_str_hash g_str_hash_traditional
+
 //! \deprecated Use g_hash_table_new_full() instead
 static inline GHashTable *
 crm_str_table_new(void)
```
Comment (clumens) [inline include/crm/common/util_compat.h:86]: I am really liking having all the deorecated code in its own headers.  It makes it a lot more obvious what we shouldn't be using and what can go away in the future.  The only other thing that might be nice to do would be include a comment on when things are expected to be removed.

**Comment by clumens** [comment]
This looks good to me.

**Pull request merged**